opensquirrel/register_manager.py

```diff
@@ -1,24 +1,142 @@
 
 import cqasm.v3x as cqasm
-
-if TYPE_CHECKING:
-    from typing import Self
+from typing_extensions import Self
 
 
 def is_qubit_type(variable: cqasm.semantic.MultiVariable) -> bool:
     return isinstance(variable.typ, cqasm.types.Qubit) or isinstance(variable.typ, cqasm.types.QubitArray)
 
 
+def is_bit_type(variable: cqasm.semantic.MultiVariable) -> bool:
+    return isinstance(variable.typ, cqasm.types.Bit) or isinstance(variable.typ, cqasm.types.BitArray)
+
+
 @dataclass
-class QubitRange:
+class Range:
     first: int = 0
     size: int = 0
 
+    def __repr__(self) -> str:
+        return "[{}{}]".format(self.first, "" if self.size == 1 else f"..{self.first + self.size - 1}")
+
+
+class Register(ABC):
+    """Register manages a (virtual) register."""
+
+    @property
+    @abstractmethod
+    def name(self) -> str: ...
+
+    @staticmethod
+    @abstractmethod
+    def is_of_type(variable: cqasm.semantic.MultiVariable) -> bool: ...
+
+    def __init__(
+        self,
+        register_size: int,
+        variable_name_to_range: dict[str, Range] | None = None,
+        index_to_variable_name: dict[int, str] | None = None,
+    ) -> None:
+        self.register_size: int = register_size
+        self.variable_name_to_range: dict[str, Range] = variable_name_to_range or dict()
+        self.index_to_variable_name: dict[int, str] = index_to_variable_name or dict()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Register):
+            return False
+        return (
+            self.register_size == other.register_size
+            and self.variable_name_to_range == other.variable_name_to_range
+            and self.index_to_variable_name == other.index_to_variable_name
+        )
+
+    def get_variable_name(self, index: int) -> str:
+        """Get the variable name at `index`."""
+        return self.index_to_variable_name[index]
+
+    def get_range(self, variable_name: str) -> Range:
+        """Get the Range for a `variable_name`."""
+        return self.variable_name_to_range[variable_name]
+
+    def get_index(self, variable_name: str, sub_index: int) -> int:
+        """Get the Index for a given `subIndex` of a `variable_name`."""
+        return self.variable_name_to_range[variable_name].first + sub_index
+
+    def __repr__(self) -> str:
+        entries: str = ""
+        first: bool = True
+        for variable_name, register_range in self.variable_name_to_range.items():
+            entries += "{}{}: {}".format("" if first else ", ", variable_name, register_range)
+            first = False
+        return f"{{ {entries} }}"
+
+    def size(self) -> int:
+        return self.register_size
+
+    @staticmethod
+    def _parse_ast_string(string: str) -> str:
+        # FIXME: libqasm should return bytes, not the __repr__ of a bytes object ("b'q'")
+        return string[2:-1]
+
+    @classmethod
+    def from_ast(cls, ast: cqasm.semantic.Program) -> Self:
+        register_size: int = 0
```

Comment:
Is it necessary to initialize the (local) variable `register_size` here? It is overwritten in line 93.

Reply:
Done!